x/superfluid/keeper/stake.go

```diff
@@ -656,3 +659,198 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+	exitCoins, err := k.forceUnlockAndExitBalancerPool(ctx, sender, poolIdLeaving, lock, sharesToStake, sdk.NewCoins())
+	if err != nil {
+		return sdk.ZeroInt(), sdk.ZeroDec(), err
+	}
+
+	totalAmtConverted, shares, err = k.convertGammSharesToOsmoAndStake(ctx, sender, valAddr, poolIdLeaving, exitCoins, minAmtToStake)
+	if err != nil {
+		return sdk.ZeroInt(), sdk.ZeroDec(), err
+	}
+
+	return totalAmtConverted, shares, nil
+}
+
+// convertUnlockedToStake converts unlocked lock shares to osmo and stakes.
+func (k Keeper) convertUnlockedToStake(ctx sdk.Context, sender sdk.AccAddress, valAddr string,
+	sharesToStake sdk.Coin, minAmtToStake sdk.Int) (totalAmtConverted sdk.Int, shares sdk.Dec, err error) {
+	// do basic validation before converting
+	poolIdLeaving, err := k.validateUnbondConvertAndStake(ctx, sharesToStake)
+	if err != nil {
+		return sdk.ZeroInt(), sdk.ZeroDec(), err
+	}
+
+	// Exit the balancer pool position.
+	// we exit with min token out amount zero since we are checking min amount designated to stake later on anyways.
+	exitCoins, err := k.gk.ExitPool(ctx, sender, poolIdLeaving, sharesToStake.Amount, sdk.NewCoins())
+	if err != nil {
+		return sdk.ZeroInt(), sdk.ZeroDec(), err
+	}
+
+	totalAmtConverted, shares, err = k.convertGammSharesToOsmoAndStake(ctx, sender, valAddr, poolIdLeaving, exitCoins, minAmtToStake)
+	if err != nil {
+		return sdk.ZeroInt(), sdk.ZeroDec(), err
+	}
+
+	return totalAmtConverted, shares, nil
+}
+
+// validateUnbondConvertAndStake validates
+// - given token has gamm share prefix
+// - gamm share denom is in correct format
+func (k Keeper) validateUnbondConvertAndStake(ctx sdk.Context, sharesToStake sdk.Coin) (poolIdLeaving uint64, err error) {
+	// Defense in depth, ensuring the sharesToMigrate contains gamm pool share prefix.
+	if !strings.HasPrefix(sharesToStake.Denom, gammtypes.GAMMTokenPrefix) {
+		return 0, types.SharesToMigrateDenomPrefixError{Denom: sharesToStake.Denom, ExpectedDenomPrefix: gammtypes.GAMMTokenPrefix}
+	}
+
+	// Get the balancer poolId by parsing the gamm share denom.
+	poolIdLeaving, err = gammtypes.GetPoolIdFromShareDenom(sharesToStake.Denom)
+	if err != nil {
+		return 0, err
+	}
+
+	return poolIdLeaving, nil
+}
+
+// convertGammSharesToOsmoAndStake converts given gamm shares to osmo by swapping in the given pool
+// then stakes it to the designated validator.
+// minAmtToStake works as slippage bound, and would error if total amount being staked is less than min amount to stake.
+// If valAddr is empty, we attempt to get staking preference from valset-pref module and stake to the given validator.
+func (k Keeper) convertGammSharesToOsmoAndStake(
+	ctx sdk.Context,
+	sender sdk.AccAddress, valAddr string,
+	poolIdLeaving uint64, exitCoins sdk.Coins, minAmtToStake sdk.Int,
+) (totalAmtCoverted sdk.Int, shares sdk.Dec, err error) {
+	var nonOsmoCoins sdk.Coins
+	bondDenom := k.sk.BondDenom(ctx)
+
+	// from the exit coins, separate non-bond denom and bond denom.
```

Comment:
This is a broader question about the feature: is the intention that we only support OSMO-paired pools (i.e. ones that can be superfluid staked)? It seems as though we never check for one of the denoms to be uosmo or for the pool to be superfluid enabled

I don't see any immediate issues with allowing this feature on all balancer pools, but still asking because it's a deviation from earlier discussions with @czarcas7ic

Reply:
I think it gets indirectly checked when we attempt to swap the denom for osmo. If the pool doesn't contain osmo, this swap will fail.

Reply:
yes, ^ this is the case